design/pool-configuration.md

```diff
@@ -0,0 +1,386 @@
+* Provide a list of the possible configuration enhancements, and to highlight dependencies and conflicts between each other (if any).
+* Highlight the cases where the solution must be BGP or L2 specific
+
+### Non-Goals
+
+* Covering only a specific use case.
+* Providing a L2 only or BGP only solution where the problem might be expressed in a more generic way.
+
+## Proposal
+
+For each ip of an address pool, we need to address three main points:
+
+* Which nodes must be advertised as destinations of that ip
+* Where / to whom the ip address must be advertised to
+* How to allocate that ip to a given service
+
+The general idea of this proposal is that these degrees of configuration are all related to the AddressPool object, and
+that these configurations are independent between each other.
+
+### Node / Service IP mapping
+
+The most common request is being able to associate a given pool to a subset of nodes (which translates to selectively announce a subset of speakers to announce the ip from).
+
+This covers use cases where different set of nodes are exposed to different subnets, and the loadbalancer IP can be reached only via a subset of the nodes.
+
+[<img src="./images/ip_per_node.svg" width="400"/>](./images/ip_per_node.svg)
+
+This was raised in the following issues / PRs:
+
+* [Issue: L2 Mode on k8s workers on different L2 segments](https://github.com/metallb/metallb/issues/511)
+* [PR: Address Pool with node selector filtering](https://github.com/metallb/metallb/pull/502)
+* [Issue: Layer 2: impossible to use MetalLB on a subset of Kubernetes nodes](https://github.com/metallb/metallb/issues/302)
+* [Issue: Layer2 Different subnets(label for LB, check for interface](https://github.com/metallb/metallb/issues/737)
+* [Issue: Look into a generic pool <> node affinity mechanism](https://github.com/metallb/metallb/issues/630)
+* [PR: MetalLB speakers working on a subset of nodes in layer2](https://github.com/metallb/metallb/pull/613)
+* [Issue: Annotate nodes with address pool/weighted address pools](https://github.com/metallb/metallb/issues/941)
+
+The following issues also fall in the same "speaker announcement logic" category:
+
+Service with external backends
+
+* [Issue: LoadBalancer service is not accessible using external IP, when endpoints (with a pod inside the cluster) is created manually](https://github.com/metallb/metallb/issues/717)
+* [Issue: Announce route for nodeless endpoints](https://github.com/metallb/metallb/issues/646)
+
+### IP Announcement Targets
+
+Currently, an IP assigned to a Service is advertised to all the possible destinations from all the nodes.
+The only exception is the BGP peer [node selector](https://metallb.universe.tf/configuration/#limiting-peers-to-certain-nodes).
+
+Here, we suggest to extend the configuration logic with a protocol dependent way to select the target(s) of the announcement.
```

Comment:
independent?

Reply:
Do you know why we can't use the BGP method with Layer2?

Reply:
Not sure I understand the question here.
This part is about choosing the targets of the advertisement, which (I believe) is protocol dependent because of the different nature of the "targets" in the BGP and the L2 context.

The part above (node / addresspool mapping) is meant to be protocol independent, and as such the one below (pool / service association)

Reply:
Yeah and reading over this, I agree. I thought you were proposing doing something with L2 as well (selectively choosing interfaces) but you say below it wasn't included in this proposal. Do we think we will ever need or want to do this?